enhancements/sig-architecture/224-event-based-manifestwork/README.md

```diff
@@ -0,0 +1,788 @@
+# Event Based Manifestwork
+
+## Release Signoff Checklist
+
+- [x] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [website](https://github.com/open-cluster-management-io/open-cluster-management-io.github.io/)
+
+## Summary
+
+This proposal optimizes scaling and performance by treating manifestworks as event messages, reducing excessive storage on the hub cluster. This allows managed cluster agents to subscribe to relevant manifestworks, eliminating the need for extensive hub storage. The proposed changes also enable diverse architectures to consume ManifestWorks, while providing a client interface for seamless adoption of the event-based mechanism.
+
+## Motivation
+
+Manifestworks are used to deliver manifests from a hub cluster to managed clusters. As manifestwork adoption and managed cluster numbers increase, we encounter scaling issues due to excessive manifestwork creation on the hub. This adversely affects the performance of the hub apiserver. This proposal is to explore treating manifestworks as event messages instead of storing them in etcd.
+
+By considering each cluster namespace as a separate topic or channel, managed cluster agents can subscribe to relevant manifestworks. This approach eliminates the need to store manifestworks on the hub cluster, as they are primarily consumed by controllers like the addon manager or manifestworkreplicaset controller. Additionally, this change allows other architectures, such as non-Kubernetes restful services or databases, to consume ManifestWorks.
+
+Moreover, with the adoption of the event-based mechanism, ManifestWorks become highly versatile and applicable in IoT or edge computing scenarios. In IoT setups, where resource constraints and intermittent connectivity are common, the event-driven approach allows ManifestWorks to be efficiently distributed and applied to edge clusters. This enhances the overall flexibility and adaptability of ManifestWorks, extending their utility beyond traditional Kubernetes environments.
+
+To facilitate easy adoption of the event-based mechanism, a client interface should be provided, enabling existing work agents or consumers to transition without extensive code changes.
+
+### Goals
+
+- Establish a standardized event protocol for delivering manifestworks and returning status information.
+- Define a resync mechanism to handle failures and ensure synchronization.
+- Develop a client interface that facilitates easy integration for consumers.
+
+### Non-Goals
+
+- Define the controlplane to act as a consumer of the protocol.
+
```

Comment:
There are several other constraints we are imposing. These need to be clarified because they are not the only way of doing this
Are we messaging framework agnostic or are we talking about working with MQTT Broker.
We need a topic per managed cluster specifically for manifestwork (there could be other topics for the same cluster)

Reply:
the message needs to follow the cloudevents. we do not limit to MQTT only. it can be other transport layers. But we take MQTT as an example or initial implementation now.

maybe add storage and persistence requirement as Non-Goals. WDYT @morvencao

Reply:
Added a new section about Source and Broker definition, also updated the non-goals for the Broker storage and persistence, please take another look and give your comments.

Reply:
looks good. Thank you Longlong